---
title: "WhereIsMyPeak"
author: "Thomas Schwarzl"
date: "24 March 2015"
output: html_document
---


Where is my peak? 

Genomic annotation with priorisation what gene types you are interested. This package was inspired
by ChIPseeker, however overcomes limited functionlity in Genomic Annotation.

This is currently under development.

```{r, eval=FALSE} 

library(WhereIsMyPeak)
library(GenomicFeatures)
library(BiocParallel)
library(biomaRt)

data("peaks", package="WhereIsMyPeak")
   
# ENSEMBL GENES 75
txdb <- makeTranscriptDbFromBiomart( biomart         = "ENSEMBL_MART_ENSEMBL",
                                     dataset         = "hsapiens_gene_ensembl",
                                     transcript_ids  = NULL,
                                     filters         = "",
                                     id_prefix       = "ensembl_",
                                     host            = "feb2014.archive.ensembl.org",
                                     port            = 80,
                                     miRBaseBuild    = NA )

#listMarts(host="feb2014.archive.ensembl.org")

ensembl75 <- useMart(host="feb2014.archive.ensembl.org", biomart="ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl")        
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl75)   

mappingTable <- getBM(c("ensembl_gene_id", "gene_biotype") , filters="", values=T, uniqueRows = TRUE, mart=ensembl)  
rownames(mappingTable) <- mappingTable[,1]  

namingTable <- getBM(c("ensembl_gene_id", "external_gene_id", "description") , filters="", values=T, uniqueRows = TRUE, mart=ensembl)  
rownames(namingTable) <- namingTable[,1]
namingTable <- namingTable[,-1]

txdbParsed <- initRegions(txdb)
annotatedPeaks <- WhereIsMyPeak(peaks, txdbParsed, mappingTable=mappingTable, namingTable=namingTable)
```


```{r} 

# annotatedPeaks <- lapply(peaks.filtered.list, function(x) {
#       ttt <- as.data.frame(x)
#       ttt[,1] <- gsub("chr", "", ttt[,1])
#       ttt <- ttt[,-4]
#       ttt <- makeGRangesFromDataFrame(ttt, keep.extra.columns = T)
#       WhereIsMyPeak(ttt, txdb=txdb, mappingTable=mappingTable, namingTable=namingTable)
# })
```


If peaks were in a list and you have to remove the "chr" of the chromosome name, then you would do something like this.

```{r, eval=FALSE} 
# annotatedPeaks <- lapply(peaks.filtered.list, function(x) {
#       ttt <- as.data.frame(x)
#       ttt[,1] <- gsub("chr", "", ttt[,1])
#       ttt <- ttt[,-4]
#       ttt <- makeGRangesFromDataFrame(ttt, keep.extra.columns = T)
#       WhereIsMyPeak(ttt, txdb=txdb, mappingTable=mappingTable, namingTable=namingTable)
# })

```

To benchmark this package, do this:

```{r, eval=FALSE} 
# devtools::install_github("hadley/lineprof")
# library(lineprof)
# x <- lineprof(WhereIsMyPeak(peaks, txdb=txdb, mappingTable=mappingTable, namingTable=namingTable),)
# shine(x)
```


```{r} 

# peaks <- makeGRangesFromDataFrame(as.data.frame(data.frame(chrom=c("3", "1"), start=c(142139167, 1),	end=c(142139183,2), strand=c("*", "+"))))

```


```{r} 
txdbParsed <- initRegions(txdb)
annotationOrder = c("Exon", "Intron")
exonAnnotationOrder = c("CDS", "5' UTR", "3' UTR")
typeOrder = c( "non_coding",
                 "known_ncrna",
                 "rRNA",
                 "snoRNA",
                 "snRNA",
                 "miRNA", 
                 "lincRNA", 
                 "Mt_rRNA", 
                 "Mt_tRNA",
                 "sense_intronic",
                 "transcribed_processed_pseudogene",
                 "transcribed_unitary_pseudogene",   
                 "transcribed_unprocessed_pseudogene",
                 "translated_processed_pseudogene",
                 "translated_unprocessed_pseudogene",
                 "unitary_pseudogene",
                 "unprocessed_pseudogene",
                 "sense_overlapping",
                 "3prime_overlapping_ncrna",   
                 "antisense",
                 "IG_C_gene",
                 "IG_D_gene",
                 "IG_J_gene",
                 "IG_V_gene",
                 "IG_C_pseudogene", 
                 "IG_J_pseudogene",
                 "IG_V_pseudogene", 
                 "polymorphic_pseudogene",
                 "misc_RNA",
                 "TEC",
                 "TR_C_gene",
                 "TR_D_gene",
                 "TR_J_gene",
                 "TR_J_pseudogene",
                 "TR_V_gene",
                 "TR_V_pseudogene",
                 "pseudogene",
                 "processed_pseudogene", 
                 "processed_transcript",
                 "LRG_gene", 
                 "protein_coding")

out <- WhereIsMyPeak(peaks,  txdbParsed, mappingTable, namingTable,
                     annotationOrder, exonAnnotationOrder, typeOrder)
```

